### PR TITLE
Add attribute get api

### PR DESCRIPTION
**Pull request by prajeeshprakash**
backup_resrc_attr utility function changes will be removed once the demo branch is updated with utkarsha's changes

### CODE REVIEW

File: resource_mgmt.c [hunk 1]
```diff
@@ -907,7 +932,6 @@ int coco_device_resource_attribute_update(coco_std_resource_attribute_info_t *re
 
   if (!coco_internal_register_other_api_ev()) {
     ec_debug_log(LOG_ERR, "coco_device_resource_attribute_update cannot be called in this sequence", NULL);
-    // TODO: set cocogatewaysdk errno
```
Comment (varun-elear) [inline resource_mgmt.c]: Why has this comment been removed?

File: resource_mgmt.c [hunk 2]
```diff
@@ -665,26 +672,43 @@ Output(s)   : coco_resource_summary_info_t *: Returns the backup of resource sum
 Description : Validates and backups resource summary struct
 *******************************************************************************/
 coco_std_resource_attribute_info_t *backup_resrc_attr(
-  coco_std_resource_attribute_info_t *resourceAttribute
-) {
+  coco_std_resource_attribute_info_t *resourceAttribute) {
```
Comment (varun-elear) [inline resource_mgmt.c]: How do we update this utility as per changes made by Utkarsha?

File: resource_mgmt.c [hunk 3]
```diff
@@ -1149,3 +1173,101 @@ int32_t coco_device_send_device_cmd_status(coco_std_device_cmd_status_t *deviceC
...
+  ec_debug_log(LOG_DEBUG, "Started", NULL);
+  int i = 0;
+  int32_t status;
+  char *runningPtr = NULL;
+  char *searchCriteria = NULL;
+  int searchCriteriaLength = 0;
+
+  if (!coco_internal_register_other_api_ev()) {
+    ec_debug_log(LOG_ERR, "coco_device_resource_attribute_get cannot be called in this sequence", NULL);
+    return -1;
```
Comment (varun-elear) [inline resource_mgmt.c]: Set errno in all error paths

File: resource_mgmt.c [hunk 4]
```diff
@@ -1149,3 +1173,101 @@ int32_t coco_device_send_device_cmd_status(coco_std_device_cmd_status_t *deviceC
...
+                                           int *attributeCount) {
+  ec_debug_log(LOG_DEBUG, "Started", NULL);
+  int i = 0;
+  int32_t status;
+  char *runningPtr = NULL;
+  char *searchCriteria = NULL;
+  int searchCriteriaLength = 0;
+
+  if (!coco_internal_register_other_api_ev()) {
+    ec_debug_log(LOG_ERR, "coco_device_resource_attribute_get cannot be called in this sequence", NULL);
```
Comment (varun-elear) [inline resource_mgmt.c]: Error logs should start with `Error: `

File: resource_mgmt.c [hunk 5]
```diff
@@ -1149,3 +1173,101 @@ int32_t coco_device_send_device_cmd_status(coco_std_device_cmd_status_t *deviceC
...
+  if (NULL == getAttrParam->resourceEui) {
+    ec_debug_log(LOG_ERR, "Error: resourceEui cannot be NULL", NULL);
+    return -1;
+  }
+
+  if (0 > getAttrParamCount) {
+    ec_debug_log(LOG_ERR, "Error: Invalid getAttrParamCount value", NULL);
+    return -1;
+  }
+  /* Calculate the total search query length */
```
Comment (varun-elear) [inline resource_mgmt.c]: Add a blank line

File: resource_mgmt.c [hunk 6]
```diff
@@ -1149,3 +1173,101 @@ int32_t coco_device_send_device_cmd_status(coco_std_device_cmd_status_t *deviceC
...
+    ec_debug_log(LOG_ERR, "coco_device_resource_attribute_get cannot be called in this sequence", NULL);
+    return -1;
+  }
+
+  if (NULL == getAttrParam) {
+    ec_debug_log(LOG_ERR, "Error: getAttrParam cannot be NULL", NULL);
+    return -1;
+  }
+
+  if (NULL == getAttrParam->resourceEui) {
```
Comment (varun-elear) [inline resource_mgmt.c]: This check should be done after `if (0 > getAttrParamCount)`. Also, this check needs to be performed for all elements in the array `getAttrParam`

File: resource_mgmt.c [hunk 7]
```diff
@@ -1149,3 +1173,101 @@ int32_t coco_device_send_device_cmd_status(coco_std_device_cmd_status_t *deviceC
...
+    ec_debug_log(LOG_ERR, "Error: Invalid getAttrParamCount value", NULL);
+    return -1;
+  }
+  /* Calculate the total search query length */
+  /* searchCriteriaLength = Length of resourceEui + Length of capabilityId + Length of attributeId
+                            Search criteria default query length + ordered by query length +
+                            OR clause string length + NULL termination */
+  for (i = 0; i < getAttrParamCount; i++) {
+    searchCriteriaLength += strlen(getAttrParam[i].resourceEui) +
+                                   ec_strlen_uint(getAttrParam[i].capabilityId) +
```
Comment (varun-elear) [inline resource_mgmt.c]: Fix indentation

File: resource_mgmt.c [hunk 8]
```diff
@@ -1149,3 +1173,101 @@ int32_t coco_device_send_device_cmd_status(coco_std_device_cmd_status_t *deviceC
...
+                 getAttrParam[i].attributeId) < 0) {
+      ec_debug_log(LOG_ALERT, "Fatal: unable to form the searchCriteria, %s", ecSuicideString);
+      ec_cleanup_and_exit();
+    }
+    // Condition check for adding the 'OR' clause and the end of each query except the last query
+    if (i < getAttrParamCount - 1) {
+      strncat(runningPtr, COCO_INTRNL_ATTR_QUERY_OR_STRING, strlen(COCO_INTRNL_ATTR_QUERY_OR_STRING));
+      runningPtr += strlen(runningPtr);
+    }
+
```
Comment (varun-elear) [inline resource_mgmt.c]: This blank line is not required.

File: resource_mgmt.c [hunk 9]
```diff
@@ -1149,3 +1173,101 @@ int32_t coco_device_send_device_cmd_status(coco_std_device_cmd_status_t *deviceC
...
+  }
+  // Append the ordered by query at the of search query
+  strncat(runningPtr, COCO_INTRNL_ATTR_ORDERED_BY, COCO_INTRNL_ATTR_ORDERED_BY_LEN);
+
+  ec_debug_log(LOG_DEBUG, "SearchCriteria : %s", searchCriteria);
```
Comment (varun-elear) [inline resource_mgmt.c]: Remove this log

File: resource_mgmt.c [hunk 10]
```diff
@@ -1149,3 +1173,101 @@ int32_t coco_device_send_device_cmd_status(coco_std_device_cmd_status_t *deviceC
...
+  status = cocodb_fetch_data(COCODB_RESOURCE_CAPABILITY_ATRIB_RECORD, searchCriteria,
+                             strlen(searchCriteria), attributeCount, (void **)attributeInfo);
+
+  if (-1 == ec_deallocate(searchCriteria)) {
+    ec_debug_log(LOG_ALERT, "Fatal: Unable to deallocate searchCriteria buffer, %s",
+                 ecSuicideString);
+    ec_cleanup_and_exit();
+  }
+
+  if ((COCO_DB_SUCCESS != status) || (COCO_DB_NO_ROW == status)) {
```
Comment (varun-elear) [inline resource_mgmt.c]: If `(COCO_DB_NO_ROW == status)`, return success

File: resource_mgmt.c [hunk 11]
```diff
@@ -1149,3 +1173,101 @@ int32_t coco_device_send_device_cmd_status(coco_std_device_cmd_status_t *deviceC
...
+  /* searchCriteriaLength = Length of resourceEui + Length of capabilityId + Length of attributeId
+                            Search criteria default query length + ordered by query length +
+                            OR clause string length + NULL termination */
+  for (i = 0; i < getAttrParamCount; i++) {
+    searchCriteriaLength += strlen(getAttrParam[i].resourceEui) +
+                                   ec_strlen_uint(getAttrParam[i].capabilityId) +
+                                   ec_strlen_uint(getAttrParam[i].attributeId) +
+                                   COCO_INTRNL_GET_ATTR_SERCH_CRITERIA_LEN +
+                                   COCO_INTRNL_ATTR_ORDERED_BY_LEN +
+                                   strlen(COCO_INTRNL_ATTR_QUERY_OR_STRING) + 1;
```
Comment (varun-elear) [inline resource_mgmt.c]: Do not include `strlen(COCO_INTRNL_ATTR_QUERY_OR_STRING)` for `getAttrParamCount - 1 == i`

File: resource_mgmt.c [hunk 12]
```diff
@@ -1149,3 +1173,101 @@ int32_t coco_device_send_device_cmd_status(coco_std_device_cmd_status_t *deviceC
...
+  }
+  // Running pointer is used to concatenate the search query one after another.
+  runningPtr = searchCriteria;
+  // Loop iteration for forming the cocodb attribute record search query
+  for (i = 0; i < getAttrParamCount; i++) {
+    // Calculate the length of the first set search query
+    searchCriteriaLength = strlen(getAttrParam[i].resourceEui) +
+                           ec_strlen_uint(getAttrParam[i].capabilityId) +
+                           ec_strlen_uint(getAttrParam[i].attributeId) +
+                           COCO_INTRNL_GET_ATTR_SERCH_CRITERIA_LEN + 5;
```
Comment (varun-elear) [inline resource_mgmt.c]: Please add a comments explaining what `+5` is being used for

File: resource_mgmt.c (hunk 1 above)
Comment (prajeeshprakash) [reply resource_mgmt.c]: I will update this with error number setting

File: resource_mgmt.c (hunk 2 above)
Comment (prajeeshprakash) [reply resource_mgmt.c]: We will remove this utility function and use the utility function implemented by utkarsha

File: resource_mgmt.c [hunk 13]
```diff
@@ -1189,36 +1189,49 @@ int32_t coco_device_resource_attribute_get(coco_device_get_attr_param_t *getAttr
...
-    return -1;
-  }
-
-  if (NULL == getAttrParam->resourceEui) {
-    ec_debug_log(LOG_ERR, "Error: resourceEui cannot be NULL", NULL);
+    cocoDeviceErrno = COCO_GATEWAY_API_EINVAL;
     return -1;
   }
 
   if (0 > getAttrParamCount) {
```
Comment (varun-elear) [inline resource_mgmt.c]: `0 >=`

File: resource_mgmt.c [hunk 14]
```diff
@@ -1189,36 +1189,49 @@ int32_t coco_device_resource_attribute_get(coco_device_get_attr_param_t *getAttr
...
-                                   COCO_INTRNL_GET_ATTR_SERCH_CRITERIA_LEN +
-                                   COCO_INTRNL_ATTR_ORDERED_BY_LEN +
-                                   strlen(COCO_INTRNL_ATTR_QUERY_OR_STRING) + 1;
+                            ec_strlen_uint(getAttrParam[i].capabilityId) +
+                            ec_strlen_uint(getAttrParam[i].attributeId) +
+                            COCO_INTRNL_GET_ATTR_SERCH_CRITERIA_LEN +
+                            COCO_INTRNL_ATTR_ORDERED_BY_LEN;
+
+    if (i < getAttrParamCount - 1) {
+      searchCriteriaLength += strlen(COCO_INTRNL_ATTR_QUERY_OR_STRING) + 1;
```
Comment (varun-elear) [inline resource_mgmt.c]: Why do we need `+1` here?

File: resource_mgmt.c [hunk 15]
```diff
@@ -1232,7 +1245,7 @@ int32_t coco_device_resource_attribute_get(coco_device_get_attr_param_t *getAttr
     searchCriteriaLength = strlen(getAttrParam[i].resourceEui) +
                            ec_strlen_uint(getAttrParam[i].capabilityId) +
                            ec_strlen_uint(getAttrParam[i].attributeId) +
-                           COCO_INTRNL_GET_ATTR_SERCH_CRITERIA_LEN + 5;
+                           COCO_INTRNL_GET_ATTR_SERCH_CRITERIA_LEN + 5; // Query length for ' OR ' + NULL termination
```
Comment (varun-elear) [inline resource_mgmt.c]: We don't need length of OR for last record. Also, why not use `strlen(COCO_INTRNL_ATTR_QUERY_OR_STRING)`?

File: resource_mgmt.c [hunk 16]
```diff
@@ -1265,9 +1275,14 @@ int32_t coco_device_resource_attribute_get(coco_device_get_attr_param_t *getAttr
     ec_debug_log(LOG_ERR, "Error: Unable to fetch attribute record", NULL);
     *attributeCount = 0;
     *attributeInfo = NULL;
-    return -1;
+    if (COCO_DB_NO_ROW != status) {
+      ec_debug_log(LOG_ERR, "Error: No rows are available", NULL);
```
Comment (varun-elear) [inline resource_mgmt.c]: Update error message: `Error: Fetching attributes failed`

File: resource_mgmt.c [hunk 17]
```diff
@@ -1189,36 +1189,49 @@ int32_t coco_device_resource_attribute_get(coco_device_get_attr_param_t *getAttr
...
     searchCriteriaLength += strlen(getAttrParam[i].resourceEui) +
-                                   ec_strlen_uint(getAttrParam[i].capabilityId) +
-                                   ec_strlen_uint(getAttrParam[i].attributeId) +
-                                   COCO_INTRNL_GET_ATTR_SERCH_CRITERIA_LEN +
-                                   COCO_INTRNL_ATTR_ORDERED_BY_LEN +
-                                   strlen(COCO_INTRNL_ATTR_QUERY_OR_STRING) + 1;
+                            ec_strlen_uint(getAttrParam[i].capabilityId) +
+                            ec_strlen_uint(getAttrParam[i].attributeId) +
+                            COCO_INTRNL_GET_ATTR_SERCH_CRITERIA_LEN +
+                            COCO_INTRNL_ATTR_ORDERED_BY_LEN;
```
Comment (varun-elear) [inline resource_mgmt.c]: COCO_INTRNL_ATTR_ORDERED_BY_LEN also needs to be added only once

**Pull request merged**